R/osf_cp.R

```diff
@@ -0,0 +1,113 @@
+#'   * An [`osf_tbl_file`] with a single directory.
+#' @template verbose
+#'
+#' @return An [`osf_tbl_file`] containing the updated OSF file.
+#'
+#' @examples
+#' \dontrun{
+#' # Create an example file to upload to our example project
+#' project <- osf_create_project("Flower Data")
+#' write.csv(iris, file = "iris.csv")
+#' data_file <- osf_upload(project,"iris.csv")
+#'
+#' # Make a copy of the file in the same location. A suffix is added to avoid name conflicts
+#' data_file <- osf_cp(data_file)
+#'
+#' # Create a new directory to copy our file to
+#' data_dir <- osf_mkdir(project, "data")
+#'
+#' # Copy the file to our data directory
+#' data_file <- osf_cp(data_file, to = data_dir)
+#'
+#' # Copy directory to new component
+#' data_comp <- osf_create_component(project, title = "data", category = "data")
+#' data_dir %>%
+#'   osf_cp(to = data_comp) %>%
+#'   osf_open()
+#' }
+#'
+#' @export
+osf_cp <- function(x, to = NULL, verbose = FALSE) {
+  UseMethod("osf_cp")
+}
+
+#' @export
+osf_cp.osf_tbl_file <- function(x, to = NULL, verbose = FALSE) {
+  x <- make_single(x)
+  if (is.null(to)) to <- osf_retrieve_node(get_parent_id(x))
+  out <- .wb_file_copy(
+    x,
+    to = to,
+    action = "copy",
+    verbose = verbose
+  )
+  as_osf_tbl(out["data"], subclass = "osf_tbl_file")
+}
+
+
+#' Internal method for moving/copying files
+#' @noRd
+#' @references
+#' https://waterbutler.readthedocs.io/en/latest/api.html#actions
+.wb_file_copy <- function(x, to, action, verbose) {
+  action <- match.arg(action, c("move", "copy"))
+  conflict <- "keep"
+
+  if (inherits(to, "osf_tbl_file")) {
+    if (is_osf_file(to)) {
+      abort("If `to` is an `osf_tbl_file` it  must contain a directory, not a file.")
+    }
+
+    # verify destination is not a child of x
+    is_child_dest <- fs::path_has_parent(
+      get_meta(to, "attributes", "materialized_path"),
+      get_meta(x, "attributes", "materialized_path")
+    )
+    if (is_child_dest) abort("Can't move a parent directory into its child.")
+  }
+
+  api_url <- get_meta(x, "links", "move")
+  api_path <- crul::url_parse(api_url)$path
+
+  req <- modifyList(
+    build_move_request(to),
+    list(action = action, conflict = conflict)
+  )
+
+  res <- .wb_request("post", api_path, body = req, encode = "json")
+  out <- process_response(res)
+  raise_error(out)
+
+  if (verbose) message(sprintf("Copied '%s' to '%s'.", x$name, to$name))
+
+  # retrieve osf representation of file
+  file_id <- strsplit(out$data$id, split = "/", fixed = TRUE)[[1]][2]
+  .osf_file_retrieve(file_id)
+}
+
+
+# Construct the move/copy request's body
+build_move_request <- function(x) UseMethod("build_move_request")
+
+build_move_request.osf_tbl_file <- function(x) {
+  list(
+    path = get_meta(x, "attributes", "path")
+  )
+}
+
+build_move_request.osf_tbl_node <- function(x) {
+  list(
+    path = "/",
+    resource = unclass(as_id(x)),
+    provider = "osfstorage"
+  )
+}
```

Comment:
No need to duplicate this code. We can just call the existing  internal  function`.wb_file_move()` located in `R/osf_mv.R`.

And since we're going to use that function for both moving *and* copying now it should probably be renamed... How about `.wb_file_action()`?

Reply:
Do you prefer the version in `osf_mv.R` ?  Otherwise i agree with the function name change.

```
# Construct the move/copy request's body
build_move_request <- function(x) {
  switch(class(x)[1],

    osf_tbl_node =   list(
      path = "/",
      resource = unclass(as_id(x)),
      provider = "osfstorage"
    ),

    osf_tbl_file = list(
      path = get_meta(x, "attributes", "path")
    )
  )
}
```

Reply:
Yeah, let's use that version. I changed it recently to avoid using internal S3 methods.